app/src/main/java/de/westnordost/streetcomplete/osm/surface/SurfaceUtils.kt

```diff
@@ -0,0 +1,158 @@
+package de.westnordost.streetcomplete.osm.surface
+
+import de.westnordost.streetcomplete.osm.getLastCheckDateKeys
+
+val INVALID_SURFACES = setOf(
+    "cobblestone", // https://wiki.openstreetmap.org/wiki/Tag%3Asurface%3Dcobblestone
+    "cement", // https://community.openstreetmap.org/t/mysterious-surface-cement/5158 and https://wiki.openstreetmap.org/wiki/Tag:surface%3Dconcrete
+)
+
+val SOFT_SURFACES = setOf("ground", "earth", "dirt", "grass", "sand", "mud", "ice", "salt", "snow", "woodchips")
+
+val ANYTHING_UNPAVED = SOFT_SURFACES + setOf(
+    "unpaved", "compacted", "gravel", "fine_gravel", "pebblestone", "grass_paver",
+    // this ones are not strictly supported by aliased - so this is used by commonSurface* functions
+    "earth", "mud",
+)
+
+val ANYTHING_FULLY_PAVED = setOf(
+    "paved", "asphalt", "cobblestone", "cobblestone:flattened", "sett",
+    "concrete", "concrete:plates", "paving_stones",
+    "metal", "wood", "unhewn_cobblestone",
+    // this ones are not strictly supported by aliased - so this is used by commonSurface* functions
+    "brick", "bricks", "cobblestone:flattened", "paving_stones:30",
+)
+
+val ANYTHING_PAVED = ANYTHING_FULLY_PAVED + setOf(
+    "concrete:lanes"
+)
+
+val INVALID_SURFACES_FOR_TRACKTYPES = mapOf(
+    "grade1" to ANYTHING_UNPAVED,
+    "grade2" to SOFT_SURFACES,
+    "grade3" to ANYTHING_FULLY_PAVED,
+    "grade4" to ANYTHING_FULLY_PAVED,
+    "grade5" to ANYTHING_FULLY_PAVED,
+)
+
+fun isSurfaceAndTracktypeMismatching(surface: String, tracktype: String): Boolean =
+    INVALID_SURFACES_FOR_TRACKTYPES[tracktype]?.contains(surface) == true
+
+fun keysToBeRemovedOnSurfaceChange(prefix: String): Set<String> =
+    setOf(
+        "${prefix}surface:grade",
+        "${prefix}smoothness",
+        "${prefix}smoothness:date",
+        "${prefix}smoothness",
+        "${prefix}surface:colour",
+        "source:${prefix}surface") +
+    getLastCheckDateKeys("${prefix}surface") +
+    getLastCheckDateKeys("${prefix}smoothness")
+
+sealed class SurfaceInfo
+sealed class SingleSurfaceInfo : SurfaceInfo()
+
+data class SingleSurface(val surface: Surface) : SingleSurfaceInfo()
+data class SingleSurfaceWithNote(val surface: Surface, val note: String) : SingleSurfaceInfo()
+object SurfaceMissing : SingleSurfaceInfo()
+data class SurfaceMissingWithNote(val note: String) : SingleSurfaceInfo()
+data class CyclewayFootwaySurfaces(val main: Surface?, val cycleway: Surface?, val footway: Surface?) : SurfaceInfo()
+data class CyclewayFootwaySurfacesWithNote(val main: Surface?, val note: String?, val cycleway: Surface?, val cyclewayNote: String?, val footway: Surface?, val footwayNote: String?) : SurfaceInfo()
```

Comment:
While reviewing, I keep stumbling over this and the complexity that is connected with dealing with this data structure.

I am not sure if it makes sense. For the `RoadSurfaceOverlay` for example, the latter two are completely irrelevant and should not at all influence how the overlay functions.  Actually, the `SurfaceAndNote` data structure should be enough.

For the path overlay which must deal with all three of sidewalk surfaces, segregated cycleway/footway surfaces and normal surfaces, this **may** be a better fit, but on the other hand I don't quite see the benefit to replicate the parsing of all these three different taggings when there is already `LeftAndRightSidewalkSurface` and `SurfaceAndNote`. If a data structure that encompasses it all is necessary, wouldn't something like
```kotlin
data class AllThemSurfaces(main: SurfaceAndNote?, sidewalk: LeftAndRightSidewalkSurface?, cyclewayPart: SurfaceAndNote?, footwayPart: SurfaceAndNote?)
```
not make more sense?

Reply:
mostly changed, and yes it is much better now